### ISSUE TITLE
Non-inheritable properties on an object

### ISSUE THREAD

**Issue by fatcerberus** [svaarala/duktape#1018]
Is it possible for C code to define non-inheritable properties on an object, like the "internal slots" described in the Ecmascript specifications?  While C code can easily create internal properties using the `\xff` prefix, these are inherited through the prototype chain like any other property.

**Comment by svaarala** [comment]
Not at the moment but this wouldn't be difficult to implement I think - there's space in the property attributes field for an additional bit, and property lookup code could be aware of whether we're in the original object or already walked up the prototype chain.

Another approach would be to reserve a certain internal property string subset which wouldn't be inherited. But to me it'd make more sense for it to be controlled by a property attribute and `duk_def_prop()`.

**Comment by fatcerberus** [comment]
Alternatively, there could be a variant of `duk_get_prop_xxx()` that only looks at own properties.  Currently one would have to use `duk_enum()` for that, which is a bit heavy just to get at single properties.

**Comment by svaarala** [comment]
Yeah, there are internal methods to get own properties only though not very conveniently (it's only rarely needed).

There are also other property access use cases which aren't supported in the API now, for example:
- Access object properties without following Proxy - important if you want to edit/inspect a Proxy object from C code or avoid side effects.
- Access properties without triggering any side effects such as getters. Important for code dealing with errors safely, for instance.

I'd prefer to avoid adding a bunch of APIs for each circumstance, but instead maybe extend the API with a set of extensible property access functions which take behavioral flags for controlling own/inherited properties, access to internal properties, triggering Proxy behavior, triggering getters, etc. Something like:

```
/* Follow Proxy properties but don't invoke getters, don't follow prototype chain, etc. */
duk_get_prop_ext(ctx, -3, DUK_PROP_FOLLOW_PROXY);
```

It's good that the default set of property access functions map 1:1 to Ecmascript though, and are not saddled with a (typically unneeded) flags field.

**Comment by fatcerberus** [comment]
> It's good that the default set of property access functions map 1:1 to Ecmascript though, and are not saddled with a (typically unneeded) flags field.

Agree, and in the vast majority of cases that's all you ever need.  Sometimes though... :)
